### DISCUSSION TITLE
ZFS Copying files

### DISCUSSION THREAD

**Discussion by mrpepik** [openzfs/zfs#16902] [Q&A]
Hi Everyone,

I am looking at testing the OpenZFS on MacOS.  I am looking for a way to clone a 25+G file fast.  I could use AFPS if the file would be local the MacOS, but I want to host the files on a SAN and have the CI Process, so a fast clone of a target File to a "linked clone" so the CI pipeline doesn't have to wait 3+ minutes for the file to copied.

I see in `https://docs.oracle.com/en/operating-systems/solaris/oracle-solaris/11.4/manage-zfs/copying-zfs-files.html?source=%3Aow%3Alp%3Acpo%3A%3A` that ZFS has a `cp -z` command which just copies the metadata of the file, which is what the APFS `cp -c` command does.  For reference, here is a link to the APFS documentation `https://eclecticlight.co/2023/04/28/apfs-hard-links-symlinks-aliases-and-clone-files-a-summary/`

Does OpenZFS include a command that I can do this with?  OR do I need to install a full ZFS SAN and run my commands there?

Thanks in advance for any help.

MrPepik

**Comment by amotin** [comment]
ZFS does not include any `cp` commands itself.  It merely implements a set of generic file system methods allowed by specific OS, such as read, write, delete, etc.  OpenZFS got block cloning support between versions 2.2 and 2.3, but how it is to be called depends on specific OS.  I can not say anything about MacOS, since it is still not in the main repo, but for both Linux and FreeBSD there OpenZFS implements `copy_file_range()` and some other lately added methods, allowing OS tools like `cp` to clone the files faster and with much smaller space usage.  Considering respective functionality is implemented on MacOS for APFS, I tend to think it is supported by kernel.  After that it is a question to MacOS OpenZFS port, whether it includes the necessary glue.

**Comment by mrpepik** [comment]
Alexander,

Happy Holidays and thank you for the fast response.  I appreciate the
information and will stand up a ZFS system and test the copying from both
Linux and MacOS.

Thanks again
